File: backend-ruby/app/controllers/api/v1/xep_cho_controller.rb
class Api::V1::XepChoController < ApplicationController
  before_action :set_xep_cho, only: [:show, :destroy]

  def index
    records = XepCho
              .includes(:dang_ky_thi, :ky_thi, :phong_thi)
              .order(XepChoID: :asc)

    render json: records.as_json(
      include: {
        ky_thi: {
          only: [:KyThiID, :MaKyThi, :TenKyThi, :TrangThai]
        },
        phong_thi: {
          only: [:PhongThiID, :MaPhong, :TenPhong, :ToaNha, :Tang]
        },
        dang_ky_thi: {
          only: [:DangKyThiID, :SoBaoDanh],
          include: {
            sinh_vien: {
              only: [:SinhVienID, :MaSinhVien, :HoTen, :Lop]
            }
          }
        }
      }
    )
  end

  def show
    render json: @xep_cho.as_json(
      include: {
        ky_thi: {
          only: [:KyThiID, :MaKyThi, :TenKyThi, :TrangThai]
        },
        phong_thi: {
          only: [:PhongThiID, :MaPhong, :TenPhong, :ToaNha, :Tang]
        },
        dang_ky_thi: {
          only: [:DangKyThiID, :SoBaoDanh],
          include: {
            sinh_vien: {
              only: [:SinhVienID, :MaSinhVien, :HoTen, :Lop]
            }
          }
        }
      }
    )
  end

  def create
    phan_phong = PhanPhong.find_by!(DangKyThiID: xep_cho_params[:DangKyThiID])

    record = XepCho.new(xep_cho_params)
    record.KyThiID = phan_phong.KyThiID
    record.PhongThiID = phan_phong.PhongThiID

    if record.save
      render json: record, status: :created
    else
      render json: { errors: record.errors.full_messages }, status: :unprocessable_entity
    end
  rescue ActiveRecord::RecordNotFound
    render json: { error: "Room assignment not found for this registration." }, status: :not_found
  rescue ActiveRecord::RecordNotUnique
    render json: { error: "Duplicate seat or this registration already has a seat." }, status: :conflict
  rescue ActiveRecord::InvalidForeignKey => e
    render json: { error: e.message }, status: :conflict
  end

  def destroy
    @xep_cho.destroy
    render json: { message: "Deleted successfully." }
  rescue ActiveRecord::InvalidForeignKey
    render json: { error: "Cannot delete this seat assignment because it is being used." }, status: :conflict
  end

  private

  def set_xep_cho
    @xep_cho = XepCho.find(params[:id])
  end

  def xep_cho_params
    params.require(:xep_cho).permit(
      :DangKyThiID,
      :SoCho,
      :Hang,
      :Cot
    )
  end
end

File: backend-csharp/Controllers/XepChoController.cs
using System.Text.Json;
using backend_csharp.Data;
using backend_csharp.Models;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;

namespace backend_csharp.Controllers;

[Route("api/v1/xep_cho")]
public class XepChoController(ExamDbContext dbContext) : ApiControllerBase
{
    [HttpGet]
    public async Task<IActionResult> Index()
    {
        List<XepCho> records = await QueryWithIncludes().OrderBy(record => record.XepChoID).ToListAsync();
        return Ok(records);
    }

    [HttpGet("{id:int}")]
    public async Task<IActionResult> Show(int id)
    {
        XepCho? record = await QueryWithIncludes().FirstOrDefaultAsync(item => item.XepChoID == id);
        return record is null ? NotFound(Error("Seat assignment not found.")) : Ok(record);
    }

    [HttpPost]
    public async Task<IActionResult> Create([FromBody] JsonElement body)
    {
        XepCho? record = ReadBody<XepCho>(body, "xep_cho");
        if (record is null)
        {
            return UnprocessableEntity(Errors("Invalid seat assignment payload."));
        }

        PhanPhong? phanPhong = await dbContext.PhanPhong.FirstOrDefaultAsync(item => item.DangKyThiID == record.DangKyThiID);
        if (phanPhong is null)
        {
            return NotFound(Error("Room assignment not found for this registration."));
        }

        record.KyThiID = phanPhong.KyThiID;
        record.PhongThiID = phanPhong.PhongThiID;
        dbContext.XepCho.Add(record);

        try
        {
            await dbContext.SaveChangesAsync();
            return CreatedAtAction(nameof(Show), new { id = record.XepChoID }, record);
        }
        catch (DbUpdateException exception) when (IsDatabaseConflict(exception))
        {
            return Conflict(Error("Duplicate seat or this registration already has a seat."));
        }
    }

    [HttpDelete("{id:int}")]
    public async Task<IActionResult> Destroy(int id)
    {
        XepCho? record = await dbContext.XepCho.FindAsync(id);
        if (record is null)
        {
            return NotFound(Error("Seat assignment not found."));
        }

        dbContext.XepCho.Remove(record);
        try
        {
            await dbContext.SaveChangesAsync();
            return Ok(new { message = "Deleted successfully." });
        }
        catch (DbUpdateException)
        {
            return Conflict(Error("Cannot delete this seat assignment because it is being used."));
        }
    }

    private IQueryable<XepCho> QueryWithIncludes()
    {
        return dbContext.XepCho
            .Include(record => record.KyThi)
            .Include(record => record.PhongThi)
            .Include(record => record.DangKyThi)
                .ThenInclude(record => record!.SinhVien);
    }
}
